### ISSUE TITLE
Negative indices do not work with all IList / IList<T> implementations, notably the .Values property of System.Management.Automation.OrderedHashtable

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#19167]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Being able to use _negative_ indices to refer to elements from the _end_ of an array / list is a useful feature, but it isn't currently supported in all cases:

```powershell
# Note: Same applies to [ordered] too.
$oht = [System.Management.Automation.OrderedHashtable] [ordered] @{ foo = 'one'; bar = 'two'; baz = 'three' }

$oht.Values[2]
'---'
$oht.Values[-1]  # should be equivalent, but currently isn't
```

Note: 

* `[System.Management.Automation.OrderedHashtable]` apparently uses an `[ordered]`  instance behind the scenes, because the type of the `.Values` property is `System.Collections.Specialized.OrderedDictionary+OrderedDictionaryKeyValueCollection`

* [Update for clarification]: Negative indices should work for all types that implement the `IList` and/or `IList<T>` interfaces - see [below](https://github.com/PowerShell/PowerShell/issues/19167#issuecomment-1435689691).

### Expected behavior

```console
three
---
three
```

### Actual behavior

```console
three
---
```

That is, `[-1]` yielded no result.

### Environment data

```powershell
PowerShell Core 7.4.0-preview.1
```

**Comment by 237dmitry** [comment]
Works if `$($oht.Values)[-1]`

**Comment by mklement0** [comment]
@237dmitry, yes you can use `@(...)` (conceptually better than `$(...)` in the context of indexing)  as a _workaround_ (forcing _enumeration_ and _re-collecting_ creates an `[object[]]` array, where negative indexing _does_ work), but I hope it is clear that there is indeed a problem.

It would help if you clarified - up front - that what you're offering is indeed just a _workaround_, as opposed to disputing the issue at hand.

**Comment by aguzev** [comment]
Is it about replacing the indexing behavior defined for the container `$oht.Values` with one implemented by PowerShell?

Wouldn't it be confusing?

What about unusual collections which might have no concept of the last element?

**Comment by mklement0** [comment]
@aguzev, it should apply to all types that implement `IList` and/or `IList<T>` - just like it already works for `Array`, `System.Collections.ArrayList` and `System.Collections.Generic.List<T>`, for instance.

Such types are guaranteed to have a `.Count` property and support _positional_ (integer-based) indexing via the parameterized `this[int index]` property.

**Comment by aguzev** [comment]
1. For the cases of a weird `IList` implementation with special treatment for negative indices or a weird `ICollection` implementation with meaningless `Count`, the user should have a choice.
2. In C# negative index usually throws an exception. The language has special syntax `^1` supported by `Index` and `Range` types. [Ranges feature of C#](https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-8.0/ranges)

**Comment by mklement0** [comment]
@aguzev,

* Re 1:

I didn't mention `ICollection`, because it doesn't support indexing - it is irrelevant to this discussion.

Using negative indices - as provided _by PowerShell_ - is a long-established convenience, that works on the common types I've mentioned before (`@('one', 'two')[-1]`, `([System.Collections.ArrayList] @('one', 'two'))[-1]`, `([System.Collections.Generic.List[object]] @('one', 'two'))[-1]`.

What these types have in common is implementing `IList` / `IList<T>`, so to me it makes sense to _generalize_ this feature, so as not to create a confusing inconsistency.

You always have the option _not_ to use negative indices, and in - presumably quite rare cases - where the type itself implements support for negative indices with _different_ semantics, you can call the underlying parameterized property directly with `.get_Item()` e.g.:

```
# This bypasses PowerShell's negative-index support and passes the number directly to the underlying type ([Array])
# In this case, the predictable result is an error: "Index was outside the bounds of the array."
@('one', 'two').get_Item(-1)
```

* Re 2.

C# isn't relevant to this discussion, but - as an aside that also isn't relevant to this discussion - there has been talk about implementing ranges similarly in PowerShell:
 * #7940

**Comment by aguzev** [comment]
Mentioned `ICollection` because `IList` does not define `Count`. `IList` inherits `Count` from `ICollection`.

**Comment by aguzev** [comment]
Mentioned `ICollection` because `IList` does not define `Count`. `IList` inherits `Count` from `ICollection`.

**Comment by mklement0** [comment]
@aguzev, yes, `Count` is inherited. Whether any given member of `IList` is an original member of inherited is likewise irrelevant to this discussion.

**Comment by mklement0** [comment]
And, to address the "weird `ICollection` implementation" argument:

The [prescribed behavior](https://learn.microsoft.com/en-US/dotnet/api/system.collections.icollection.count?view=net-7.0#system-collections-icollection-count)  of the `.Count` property is:

> Gets the number of elements contained in the [ICollection](https://learn.microsoft.com/en-US/dotnet/api/system.collections.icollection?view=net-7.0).

This unambiguously specifies what `.Count` _should_ report.
I don't think we need to worry about rogue implementations deviating from this specification.

**Comment by aguzev** [comment]
I am happy to see `ICollection` no longer being irrelevant. Good luck!

**Comment by mklement0** [comment]
Yes, I got myself confused over the phrase "`ICollection` implementation" given that - due to the inheritance relationship - a type implemeting `IList` _implicitly_ also implements `ICollection`, and I hope we can agree that it is the prescribed meaning of the `IList` members - whether inherited from other interfaces or not - that matters.

Since I've hidden my previous comments, let me summarize here:

The support for negative indices should apply to all types that implement `IList` and/or `IList<T>` - just like it already works for `Array`, `System.Collections.ArrayList` and `System.Collections.Generic.List<T>`, for instance.

Such types are guaranteed to have a `.Count` property and support _positional_ (integer-based) indexing via the parameterized `this[int index]` property, so PowerShell can translate `[-1]` into `[<instance>.Count-1]`, for instance.

The meaning of these interface members is well-defined:

The [prescribed behavior](https://learn.microsoft.com/en-US/dotnet/api/system.collections.icollection.count?view=net-7.0#system-collections-icollection-count)  of the `.Count` property is:

> Gets the number of elements contained in the [ICollection](https://learn.microsoft.com/en-US/dotnet/api/system.collections.icollection?view=net-7.0).

The [prescribed behavior](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ilist.item?view=net-7.0)  of the `.Item[]` property is (emphasis added):

> The _zero-based_ index of the element to get or set.

(While it technically doesn't mention that the indices should be _sequential_, I think that is commonly understood, and even then PowerShell's use of negative indices wouldn't interfere. [`IList<T>` makes this less ambiguous](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.item?view=net-7.0#system-collections-generic-list-1-item(system-int32)) by mentioning that an exception can occur when "index is equal to or greater than Count.")

In short:

* `IList`'s behavior is well-defined, and we needn't worry about rogue implementations.
* PowerShell's support for negative indices (as partially already implemented) wouldn't interfere with this  behavior and _complement_ it with additional, long-established functionality.

As an aside: _Another_ aspect of PowerShell's enhanced index support _does_ already work with the types mentioned in the initial post: support for _multiple_ indices:

```powershell
$oht = [System.Management.Automation.OrderedHashtable] [ordered] @{ foo = 'one'; bar = 'two'; baz = 'three' }
$oht.Values[2, 0] # -> 'three', 'one'
```

As noted, I've hidden my previous comments, and I invite you to hide yours too, so that we minimize the distraction to future readers.
